executor/memtable_reader.go

```diff
@@ -997,3 +997,117 @@ func (e *hotRegionsHistoryRetriver) getHotRegionRowWithSchemaInfo(
 	}
 	return row, nil
 }
+
+type tikvRegionPeersRetriever struct {
+	dummyCloser
+	extractor *plannercore.TikvRegionPeersExtractor
+	retrieved bool
+}
+
+func (e *tikvRegionPeersRetriever) retrieve(ctx context.Context, sctx sessionctx.Context) ([][]types.Datum, error) {
+	if e.extractor.SkipRequest || e.retrieved {
+		return nil, nil
+	}
+	e.retrieved = true
+	tikvStore, ok := sctx.GetStore().(helper.Storage)
+	if !ok {
+		return nil, errors.New("Information about hot region can be gotten only when the storage is TiKV")
+	}
+	tikvHelper := &helper.Helper{
+		Store:       tikvStore,
+		RegionCache: tikvStore.GetRegionCache(),
+	}
+
+	if len(e.extractor.StoreIDs) == 0 && len(e.extractor.RegionIDs) == 0 {
+		regionsInfo, err := tikvHelper.GetRegionsInfo()
+		if err != nil {
+			return nil, err
+		}
+		return e.packTiKVRegionPeersRows(regionsInfo.Regions)
+	}
+
+	var regionsInfo, regionsInfoByStoreID, regionsInfoByRegionID []helper.RegionInfo
```

Comment:
I think we don't need `regionsInfoByStoreID` and `regionsInfoByRegionID` at all. You can only keep `regionsInfo` to collect all in one.

1. `GetStoreRegionsInfo` to put all regions into `regionsInfo`.
2. Check `e.extractor.RegionIDs` if needed to exclude the region IDs we already have at step 1.
3. `GetRegionInfoByID` to put all regions into `regionsInfo`.
4. `return packTiKVRegionPeersRows(regionsInfo)`.

Reply:
got, thxs a lot.

Reply:
new commit is there, retain `regionsInfoByStoreID` and use `regionMap` to keep the index in `regionsInfoByStoreID` to fast exclude.